redbot/core/commands/converter.py

```diff
@@ -162,140 +162,175 @@ class APIToken(discord.ext.commands.Converter):
         return result
 
 
-class DictConverter(dpy_commands.Converter):
-    """
-    Converts pairs of space seperated values to a dict
-    """
+# Below this line are a lot of lies for mypy about things that *end up* correct when
+# These are used for command conversion purposes. Please refer to the portion
+# which is *not* for type checking for the actual implementation
+# and ensure the lies stay correct for how the object should look as a typehint
 
-    def __init__(self, *expected_keys: str, delims: Optional[List[str]] = None):
-        self.expected_keys = expected_keys
-        self.delims = delims or [" "]
-        self.pattern = re.compile(r"|".join(re.escape(d) for d in self.delims))
+if TYPE_CHECKING:
+    DictConverter = Dict[str, str]
+else:
 
-    async def convert(self, ctx: "Context", argument: str) -> Dict[str, str]:
+    class DictConverter(dpy_commands.Converter):
+        """
+        Converts pairs of space seperated values to a dict
+        """
 
-        ret: Dict[str, str] = {}
-        args = self.pattern.split(argument)
+        def __init__(self, *expected_keys: str, delims: Optional[List[str]] = None):
+            self.expected_keys = expected_keys
+            self.delims = delims or [" "]
+            self.pattern = re.compile(r"|".join(re.escape(d) for d in self.delims))
 
-        if len(args) % 2 != 0:
-            raise BadArgument()
+        async def convert(self, ctx: "Context", argument: str) -> Dict[str, str]:
 
-        iterator = iter(args)
+            ret: Dict[str, str] = {}
+            args = self.pattern.split(argument)
 
-        for key in iterator:
-            if self.expected_keys and key not in self.expected_keys:
-                raise BadArgument(_("Unexpected key {key}").format(key=key))
+            if len(args) % 2 != 0:
+                raise BadArgument()
 
-            ret[key] = next(iterator)
+            iterator = iter(args)
 
-        return ret
+            for key in iterator:
+                if self.expected_keys and key not in self.expected_keys:
+                    raise BadArgument(_("Unexpected key {key}").format(key=key))
 
+                ret[key] = next(iterator)
 
-def get_dict_converter(*expected_keys: str, delims: Optional[List[str]] = None) -> type:
-    """
-    Returns a typechecking safe `DictConverter` suitable for use with discord.py
-    """
+            return ret
 
-    class PartialMeta(type(DictConverter)):
-        __call__ = functools.partialmethod(
-            type(DictConverter).__call__, *expected_keys, delims=delims
-        )
 
-    class ValidatedConverter(DictConverter, metaclass=PartialMeta):
-        pass
+if TYPE_CHECKING:
```

Comment:
Why is this branch needed if type hints are exactly the same?

Reply:
type hints are the same, but there's a dynamic class declaration inside which mypy doesn't allow. For the sake of people using this, making it not appear dynamic since external to the function, the typing is fine.